pkg/cloud/services/provider/provider.go

```diff
@@ -109,9 +109,22 @@ func NewClient(cloud clientconfig.Cloud, caCert []byte) (*gophercloud.ProviderCl
 	}
 	err = openstack.Authenticate(provider, *opts)
 	if err != nil {
-		return nil, nil, fmt.Errorf("providerClient authentication err: %v", err)
+		return nil, nil, "", fmt.Errorf("providerClient authentication err: %v", err)
 	}
-	return provider, clientOpts, nil
+
+	// Set ProjectID for consumers of the ProviderClient
+	// This is only needed when application credentials were used, because they
+	// do not specify a `project_id` in the clientconfig.Cloud.
+	var projectID string
+	if tokenResult, ok := provider.GetAuthResult().(tokens.CreateResult); ok {
```

Comment:
@pierreprinetti Can I get your review on this dance specifically? 👆

Reply:
LGTM 👍

Reply:
I’ll just say that if you want to protect against catastrophic Gophercloud changes, you could log something when !ok

Reply:
Looking at the code in `github.com/gophercloud/gophercloud/openstack.Authenticate`, there are 4 auth scenarios that it seems to handle:

- [Identity v2 with `tokens`](https://github.com/gophercloud/gophercloud/blob/321520e3be598fd242bb31cd2021d3b88912c42e/openstack/client.go#L139)
- [Identity v3 with `ec2tokens`](https://github.com/gophercloud/gophercloud/blob/321520e3be598fd242bb31cd2021d3b88912c42e/openstack/client.go#L232)
- [Identity v3 with `oauth1`](https://github.com/gophercloud/gophercloud/blob/321520e3be598fd242bb31cd2021d3b88912c42e/openstack/client.go#L234)
- [Identity v3 with `tokens`](https://github.com/gophercloud/gophercloud/blob/321520e3be598fd242bb31cd2021d3b88912c42e/openstack/client.go#L236)

The `AuthResult` we get back is [of the type that was used for auth](https://github.com/gophercloud/gophercloud/blob/321520e3be598fd242bb31cd2021d3b88912c42e/provider_client.go#L230), but currently we only handle the `v3/tokens.CreateResult` type.

Looking at the generic [`AuthResult`](https://github.com/gophercloud/gophercloud/blob/321520e3be598fd242bb31cd2021d3b88912c42e/auth_result.go#L50) struct they even have an example of how one should handle this.

I am going to add handling for the other types of auth currently supported by Gophercloud and failure handling for unknown (new) types.

Reply:
Apologies for the very newbie question, but are these scenarios supported in CAPO?

Reply:
We can only support authentication mechanisms which can be entirely described by clouds.yaml, so I suspect that oauth at least wouldn't be implementable?

I would personally be ok with only supporting a single keystone auth mechanism as we currently do and just logging a useful error message. However, I'm also ok with handling all the response types we know about.

Reply:
We do not specify which mechanisms we support: https://cluster-api-openstack.sigs.k8s.io/clusteropenstack/configuration.html#openstack-credential

From the user perspective I would then assume that everything that can be describe in the clouds.yaml is fair game.

---

I checked if and where we can get the Project ID for the other mechanisms:

- Identity v2 with `tokens`: `result.ExtractToken().Tenant.ID`
- Identity v3 with `ec2tokens`: Seems to return a v3 `tokens.CreateResult`: https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/extensions/ec2tokens/requests.go#L291
- Identity v3 with `oauth1`: Also returns a v3 `tokens.CreateResult`: https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/extensions/oauth1/requests.go#L138

So the only additional case we would need to handle here is the Identity v2 `tokens` mechanism, which [was removed in Queens](https://docs.openstack.org/keystone/latest/contributor/http-api.html).

Reply:
I would vote for keeping it simple and documenting that we require keystone v3. Lets return an error here if we haven't got that.

Reply:
Okay :+1:

There is always the possibility about extending this in a later PR, if anyone wants the keystone v2 support.